cpp/density/RDF.h

```diff
@@ -32,53 +29,99 @@ class RDF : public util::NdHistogram
     void reset();
 
     //! Compute the RDF
+    /*! Accumulate the given points to the histogram. Accumulation is performed
+     * in parallel on thread-local copies of the data, which are reduced into
+     * the primary data arrays when the user requests outputs.
+     */
     void accumulate(const freud::locality::NeighborQuery* neighbor_query,
                     const vec3<float>* query_points, unsigned int n_query_points,
                     const freud::locality::NeighborList* nlist, freud::locality::QueryArgs qargs);
 
-    //! \internal
-    //! helper function to reduce the thread specific arrays into one array
-    void reduceRDF();
+    //! Reduce thread-local arrays onto the primary data arrays.
+    virtual void reduce();
 
-    //! Implementing pure virtual function from parent class.
-    virtual void reduce()
+    //! Return :code:`thing_to_return` after reducing.
+    template<typename T>
+    T &reduceAndReturn(T &thing_to_return)
     {
-        reduceRDF();
+        if (m_reduce == true)
+        {
+            reduce();
+        }
+        m_reduce = false;
+        return thing_to_return;
     }
 
-    //! Get a reference to the PCF array
+    //! Get the simulation box
+    const box::Box& getBox() const
+    {
+        return m_box;
+    }
+
+    //! Get the positional correlation function.
     const util::ManagedArray<float> &getRDF()
     {
-        return getPCF();
+        return reduceAndReturn(m_pcf);
     }
 
-    //! Get a reference to the r array
-    const util::ManagedArray<float> &getR();
+    //! Get the histogram of bin distances.
+    const util::ManagedArray<unsigned int> &getBinCounts()
+    {
+        return m_histogram.getBinCounts();
+    }
 
     //! Get a reference to the N_r array.
-    /*! Mathematically, m_N_r_array[i] is the average number of points
-     * contained within a ball of radius m_r_array[i]+dr/2 centered at a given
+    /*! Mathematically, m_N_r[i] is the average number of points
+     * contained within a ball of radius getBins()[i+1] centered at a given
      * query_point, averaged over all query_points.
      */
     const util::ManagedArray<float> &getNr()
     {
-        return reduceAndReturn(m_N_r_array);
+        return reduceAndReturn(m_N_r);
+    }
+
+    float getRMax() const
+    {
+        return m_r_max;
     }
 
-    unsigned int getNBins();
+    float getRMin() const
+    {
+        return m_r_min;
+    }
+
+    //! Get bin centers.
+    std::vector<float> getR()
```

Comment:
We could change the behavior specifically for RDF, but in general multi-dimensional histograms could have different numbers of bins in different dimensions, so it makes more sense to return a vector of vectors (which becomes a list of lists in Python). If we ever actually implement the feature to specify bin locations for RDFs, then I'd also expect us to enable that feature for e.g. PMFTs, and those would definitely be lists of lists. I'm not even sure if the property on the Python side should be a NumPy array, but I left it for now as a later API question when the API rewrite is happening. I'm open to ideas on how else we could do it, though.

Reply:
I want a NumPy array as the user-facing API, but it's fine to return a vector here to Cython and then convert it with `np.asarray` so long as it returns a copy and/or cannot be invalidated.

Reply:
How would you do this for PMFTs? I think there are some API questions to sort out regarding how we specify bins for multi-dimensional arrays, which should correspond naturally to how we return the objects.

Reply:
As discussed offline, we're mostly going with my proposal of returning tuples of bins. Main modification is that each set of bins will be a numpy array rather than a list, so the cast makes sense. This is already how RDF is behaving, so I'll update PMFTs and call it good.